clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -1176,18 +1178,25 @@ private ClusterAndWaitTime waitOnMetadata(String topic, Integer partition, long
                 metadata.awaitUpdate(version, remainingWaitMs);
             } catch (TimeoutException ex) {
                 // Rethrow with original maxWaitMs to prevent logging exception with remainingWaitMs
-                throw new TimeoutException(
-                        String.format("Topic %s not present in metadata after %d ms.",
-                                topic, maxWaitMs));
+                final String errorMessage = String.format("Topic %s not present in metadata after %d ms.",
+                        topic, maxWaitMs);
+                if (metadata.getError(topic) == Errors.UNKNOWN_TOPIC_OR_PARTITION) {
+                    throw new UnknownTopicOrPartitionException(errorMessage);
```

Comment:
I think we don't want to throw this exception directly, but rather still throw `TimeoutException` but with a nested `UnknownTopicOrPartitionException` (ie, as "cause" of the timeout)?

Reply:
> I think we don't want to throw this exception directly, but rather still throw `TimeoutException` but with a nested `UnknownTopicOrPartitionException` (ie, as "cause" of the timeout)?

OK. So I changed the PR title as well:)